Anlab.Mvc/Controllers/PaymentController.cs

```diff
@@ -0,0 +1,337 @@
+            }
+
+            if (order.CreatorId != CurrentUserId)
+            {
+                ErrorMessage = "You don't have access to this order.";
+                return RedirectToAction("Index", "Order");
+            }
+
+            if (order.Status != OrderStatusCodes.AwaitingPayment)
+            {
+                ErrorMessage = "You cannot Pay until the Order is ready."; 
+                return RedirectToAction("Index", "Order");
+            }
+
+            Dictionary<string, string> dictionary = SetDictionaryValues(order, order.Creator);
+
+            ViewBag.Signature = _dataSigningService.Sign(dictionary);
+            ViewBag.PaymentDictionary = dictionary;
+            ViewBag.CyberSourceUrl = _appSettings.CyberSourceUrl;
+
+            var model = new OrderReviewModel();
+            model.Order = order;
+            model.OrderDetails = order.GetOrderDetails();
+            model.TestItems = _context.TestItems
+                .Where(a => model.OrderDetails.SelectedTests.Select(s => s.Id).Contains(a.Id)).ToList();
+
+            return View(model);
+        }
+
+
+        [HttpPost]
+        [IgnoreAntiforgeryToken]
+        public ActionResult Receipt(ReceiptResponseModel response)
+        {
+            Log.ForContext("response", response, true).Information("Receipt response received");
+
+            // check signature
+            var dictionary = Request.Form.ToDictionary(x => x.Key.ToString(), x => x.Value.ToString());
+            if (!_dataSigningService.Check(dictionary, response.Signature))
+            {
+                Log.Error("Check Signature Failure");
+                ErrorMessage = "An error has occurred. Payment not processed. If you experience further problems, contact us.";
+                return RedirectToAction("Index", "Home");
+            }
+
+            //ProcessPaymentEvent(response, dictionary); //For testing locally, can enable this.
+
+            var order = _context.Orders.SingleOrDefault(a => a.Id == response.Req_Reference_Number);
+            if (order == null)
+            {
+                Log.Error("Order not found {0}", response.Req_Reference_Number);
+                ErrorMessage = "Order for payment not found. Please contact technical support.";
+                return NotFound(response.Req_Reference_Number);
+            }
+            if (order.CreatorId != CurrentUserId)
+            {
+                ErrorMessage = "You don't have access to this order.";
+                return NotFound(response.Req_Reference_Number);
+            }
+
+            var responseValid = CheckResponse(response);
+            if (!responseValid.IsValid)
+            {
+                ErrorMessage = ErrorMessage = string.Format("Errors detected: {0}", string.Join(",", responseValid.Errors));
+                return RedirectToAction("Pay", new {id = order.Id});
+            }
+
+            //Should be good,   
+            if (order.Status == OrderStatusCodes.AwaitingPayment)

```

Comment:
@jSylvestre seems weird that you are updating the order both in the payment receipt and in the provider notify.  I think you can do it in either/both w/ a check to make sure it isn't already paid, but half here and full in the provider notify is a strange state of affairs because now you can have a paid order w/o an associated payment event for a while (or forever if the provider notify doesn't work)

Reply:
@srkirkland I believe that was what @jpknoll suggested, unless I misunderstood.

Reply:
@jSylvestre Yeah, early returns, unpack the else statements.